pkg/api/file.go

```diff
@@ -0,0 +1,265 @@
+		s.Logger.Debugf("file: parse file address %s: %v", addr, err)
+		s.Logger.Error("file: parse file address")
+		jsonhttp.BadRequest(w, "invalid file address")
+		return
+	}
+
+	// read entry.
+	j := joiner.NewSimpleJoiner(s.Storer)
+	buf := bytes.NewBuffer(nil)
+	_, err = file.JoinReadAll(j, address, buf)
+	if err != nil {
+		s.Logger.Debugf("file: read entry %s: %v", addr, err)
+		s.Logger.Error("file: read entry")
+		jsonhttp.InternalServerError(w, "error reading entry")
+		return
+	}
+	e := &entry.Entry{}
+	err = e.UnmarshalBinary(buf.Bytes())
+	if err != nil {
+		s.Logger.Debugf("file: unmarshall entry %s: %v", addr, err)
+		s.Logger.Error("file: unmarshall entry")
+		jsonhttp.InternalServerError(w, "error unmarshalling entry")
+		return
+	}
+
+	// If none match header is set always send the reply as not modified
+	// TODO: when SOC comes, we need to revisit this concept
+	noneMatchEtag := r.Header.Get("If-None-Match")
+	if noneMatchEtag != "" {
+		if e.Reference().Equal(address) {
+			w.WriteHeader(http.StatusNotModified)
+			return
+		}
+	}
+
+	// Read metadata.
+	buf = bytes.NewBuffer(nil)
+	_, err = file.JoinReadAll(j, e.Metadata(), buf)
+	if err != nil {
+		s.Logger.Debugf("file: read metadata %s: %v", addr, err)
+		s.Logger.Error("file: read netadata")
+		jsonhttp.InternalServerError(w, "error reading metadata")
+		return
+	}
+	metaData := &entry.Metadata{}
+	err = json.Unmarshal(buf.Bytes(), metaData)
+	if err != nil {
+		s.Logger.Debugf("file: unmarshall metadata %s: %v", addr, err)
+		s.Logger.Error("file: unmarshall metadata")
+		jsonhttp.InternalServerError(w, "error unmarshalling metadata")
+		return
+	}
+
+	// send the file data back in the response
+	dataSize, err := j.Size(r.Context(), address)
+	if err != nil {
+		if errors.Is(err, storage.ErrNotFound) {
+			s.Logger.Debugf("file: not found %s: %v", address, err)
+			s.Logger.Error("file: not found")
+			jsonhttp.NotFound(w, "not found")
+			return
+		}
+		s.Logger.Debugf("file: invalid root chunk %s: %v", address, err)
+		s.Logger.Error("file: invalid root chunk")
+		jsonhttp.BadRequest(w, "invalid root chunk")
+		return
+	}
+
+	outBuffer := bytes.NewBuffer(nil)
+	c, err := file.JoinReadAll(j, e.Reference(), outBuffer)
+	if err != nil && c == 0 {
+		s.Logger.Debugf("file: data read %s: %v", addr, err)
+		s.Logger.Error("file: data read")
+		jsonhttp.InternalServerError(w, "error reading data")
+		return
+	}
+	w.Header().Set("ETag", fmt.Sprintf("%q", e.Reference()))
+	w.Header().Set("Content-Disposition", fmt.Sprintf("inline; filename=\"%s\"", metaData.Filename))
+	w.Header().Set("Content-Type", metaData.MimeType)
+	w.Header().Set("Content-Length", fmt.Sprintf("%d", dataSize))
+	_, _ = io.Copy(w, outBuffer)
+}
+
+// storeMeta is used to store metadata information as a whole.
+func (s *server) storeMeta(ctx context.Context, dataBytes []byte) (swarm.Address, error) {
+	dataBuf := bytes.NewBuffer(dataBytes)
+	dataReader := io.LimitReader(dataBuf, int64(len(dataBytes)))
```

Comment:
LimitReader is not needed.

Reply:
Just make sure that splitter writes aren't called with more than one chunk of data at a time...